project_template/CMakeLists.txt

```diff
@@ -0,0 +1,104 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: MIT
+
+cmake_minimum_required(VERSION 3.15)
+
+project(az_exampleshortname
+    LANGUAGES C CXX
+    VERSION 0.1.0)
+
+option(az_exampleshortname_with_docs "Build documentation" ON)
+option(BUILD_TESTING "Build tests" ON)
+
+
+
+
+
+# this is a cmake 3.15 feature that tells find_package to 
+# look at config files before module files. This makes it easier
+# to write buildsystems that work on many platforms, espicially
+# when finding libraries that have multiple build systems.
+
+
+# if needed: add ./cmake to the module path
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+
+set(CMAKE_C_STANDARD 99)
+set(CMAKE_CXX_STANDARD 17)
+
+# default to hidden symbols when building shared libraries
+# on unix systems
+set(CMAKE_C_VISIBILITY_PRESET hidden)
+set(CMAKE_CXX_VISIBILITY_PRESET hidden)
+set(CMAKE_VISIBILITY_INLINES_HIDDEN ON)
+
+set_property(DIRECTORY PROPERTY COMPILE_OPTIONS
+    $<$<COMPILE_LANG_AND_ID:C,MSVC>:/W4 /wd4201>

```

Comment:
Notably `LANGUAGES` includes `CXX`, but we're only setting `COMPILE_OPTIONS` here for `C`. Should the two be consistent?

Reply:
@CaseyCarter should this also be /Za and not /wd4201.

Reply:
/wd4201 is (a superset of) a C11 feature, and is very useful and widely supported.

/Za really not a good idea, C99 (ish) compatibility should be enforced by compiling with warnings using compilers that actually support C99, which msvc does not. /Za also prevents you from including most windows system headers.

/W4 warns for all of the .... less than ideal ... ms extensions.

Reply:
@CaseyCarter also: I'm not really sure how to handle this for c++, since technically you could compile with a c++ compiler that was different from your C compiler. But yes both should be set..

Also: the warnings for C files may well be different from the warnings for C++ files (tests).